### ISSUE TITLE
Consider clang-tidy `readability-function-cognitive-complexity` check

### ISSUE THREAD

**Issue by ChrisThrasher** [SFML/SFML#2555]
## Subject of the issue

clang-tidy `readability-function-cognitive-complexity` docs:
https://releases.llvm.org/14.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/readability-function-cognitive-complexity.html

Tentative implementation of this check:
https://github.com/SFML/SFML/compare/master...ChrisThrasher:SFML:cognitive-complexity

Talk by Sonar employee explaining the utility of this complexity metric:
https://www.youtube.com/watch?v=el9OKGrqU6o

clang-tidy has a check that implement [this paper from SonarSource](https://www.sonarsource.com/docs/CognitiveComplexity.pdf) which defines a concept of "cognitive complexity". The idea is that it is an empirical measures of how hard it is to understand all possible code paths through a function. What I love about this check is that it forces us to reckon with functions that have exceptionally complicated control flow and break it apart into multiple functions that are hopefully easier to read.

The default cognitive complexity score that clang-tidy uses is 25 but we have functions in SFML that score _110_. It's safe to say that SFML won't get under a score of 25 but I think it's possible we can lower that threshold down to somewhere around 75 and in doing so will improve some of the most complicated bits of code we currently maintain. Not only will we improve existing code but we will ensure that no new code is checked into that repo that violates this quality gate. I expect that this process will be very incremental as we lower the cognitive complexity threshold from 110 by increments of 5 or so points. 

We can potentially merge this new check in [its current state](https://github.com/SFML/SFML/compare/master...ChrisThrasher:SFML:cognitive-complexity) which requires no changes to the code. This will set a new quality floor (albeit a very low floor) such that no new functions exceed this threshold. It'd be a minor step but still potentially worthwhile.

---

Here's a simpler example of `readability-function-cognitive-complexity`'s output from a different project:

```
/Users/thrasher/Projects/fourier-draw/src/main.cpp:27:5: error: function 'main' has cognitive complexity of 35 (threshold 25) [readability-function-cognitive-complexity,-warnings-as-errors]
int main()
    ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:37:5: note: +1, including nesting penalty of 0, nesting level increased to 1
    if (!font.loadFromFile(FONT_PATH / std::filesystem::path("font.ttf")))
    ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:49:5: note: +1, including nesting penalty of 0, nesting level increased to 1
    while (window.isOpen()) {
    ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:50:9: note: +2, including nesting penalty of 1, nesting level increased to 2
        for (auto event = sf::Event(); window.pollEvent(event);) {
        ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:51:13: note: +3, including nesting penalty of 2, nesting level increased to 3
            switch (event.type) {
            ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:56:17: note: +4, including nesting penalty of 3, nesting level increased to 4
                switch (event.key.code) {
                ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:75:9: note: +2, including nesting penalty of 1, nesting level increased to 2
        if (sf::Mouse::isButtonPressed(sf::Mouse::Button::Left)) {
        ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:77:51: note: +1
            const auto is_within_x = mouse.x >= 0 && mouse.x <= int(window.getSize().x);
                                                  ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:78:51: note: +1
            const auto is_within_y = mouse.y >= 0 && mouse.y <= int(window.getSize().y);
                                                  ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:79:13: note: +3, including nesting penalty of 2, nesting level increased to 3
            if (is_within_x && is_within_y) {
            ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:79:29: note: +1
            if (is_within_x && is_within_y) {
                            ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:83:23: note: nesting level increased to 4
                    = [position, &signal, &line, &line_shadow, &x_epicycles, &y_epicycles, &frame_count]() {
                      ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:90:17: note: +4, including nesting penalty of 3, nesting level increased to 4
                if (signal.empty()) {
                ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:92:19: note: +1, nesting level increased to 4
                } else {
                  ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:94:21: note: +5, including nesting penalty of 4, nesting level increased to 5
                    if (vector.length() > 2)
                    ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:100:9: note: +2, including nesting penalty of 1, nesting level increased to 2
        if (x_epicycles.empty()) {
        ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:109:9: note: +2, including nesting penalty of 1, nesting level increased to 2
        if (frame_count % 2 == 0) {
        ^
/Users/thrasher/Projects/fourier-draw/src/main.cpp:131:9: note: +2, including nesting penalty of 1, nesting level increased to 2
        if (++frame_count == 2 * x_epicycles.size()) {
        ^
```

**Comment by Bromeon** [comment]
Complexity is a very subjective metric, and there's not always a very linear relation with code structure. 

For example, splitting up long functions into multiple can _increase_ complexity by creating non-locality, as well as uncertainty where and how many times a function is used. Similarly, refactoring too many parameters into a struct comes with extra code for struct packing and unpacking, and often makes sense only when the struct can be reused more than once. 

Devil's advocate: is this a problem SFML really suffers from? Not just from the point of view that "complex" code may exist, but that it affects core parts of the library, hurting readability and maintainability? I'm asking because I've also made the other experience where I [needed to excessively disable similar lints](https://github.com/search?q=repo%3Agodot-rust%2Fgdext+clippy%3A%3Atoo_many_arguments&type=code), to the point I consider turning them off entirely.

That said, thanks a lot for bringing up the discussion! Maybe you have already seen quite some parts of SFML that would benefit from such automation 🙂

**Comment by ChrisThrasher** [comment]
> Complexity is a very subjective metric

I want to emphasize this an objective complexity metric. I'm sure you're aware but it bears repeating. I agree complexity in general is highly subjective though. Frustratingly subjective sometimes.

> For example, splitting up long functions into multiple can increase complexity by creating non-locality, as well as uncertainty where and how many times a function is used.

This is a possibility. It's plausible we have to make exceptions for select functions. As with a few other clang-tidy checks, 100% compliance is sometimes impractical.

> is this a problem SFML really suffers from? Not just from the point of view that "complex" code may exist, but that it affects core parts of the library, hurting readability and maintainability?

When I read through the biggest functions that this check found (scores >=80), my impression was "I sure hope I never have to touch this". Fixing bugs or enhancing features in these functions is liable to be a huge burden when they contain so much state and so much control flow. I doubt most developers could easily grok them without spending significant time and effort getting to understand the code.

---

I'll admit I have relatively little experience with writing code that adheres to these complexity metric so I may sour on this idea after enough experience. I do however think it's a worthwhile experiment. I'm personally experimenting in some personal projects to get a better idea for what it entails in practice and to get better at breaking apart big functions.

**Comment by ChrisThrasher** [comment]
A useful implementation detail I just discovered is that all you need to disable this check for a given function is to add `// NOLINTNEXTLINE(readability-function-cognitive-complexity)` above the function definition. No need for a `// NOLINTBEGIN`/`// NOLINTEND` combo. Ignoring a select few offenders would let us significantly lower the cap from 110 thus increasing the value of doing this analysis in the first place.

**Comment by ChrisThrasher** [comment]
By disabling this check for [only 4 functions](https://github.com/ChrisThrasher/SFML/commit/4feb19fc83c835078c4e279357f5ef6b6bc3893c) I was about to lower the threshold to 75. That's still very high but at least the threshold is low enough that it's plausible we accidentally violate it and thus get some worthwhile feedback from it.

If we enable the [`readability-else-after-return`](https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html) check, that will help us chip away at cognitive complexity in a lot of places. The pattern of removing unnecessary `else` blocks is highlighted in the YouTube video I linked up above as an easy way to reduce cognitive complexity. [Here is what that looks like.](https://github.com/SFML/SFML/compare/master...ChrisThrasher:SFML:readability-else-after-return)

**Comment by Bromeon** [comment]
> By disabling this check for [only 4 functions](https://github.com/ChrisThrasher/SFML/commit/4feb19fc83c835078c4e279357f5ef6b6bc3893c) I was about to lower the threshold to 75.

Should we maybe exclude examples from this check? I would argue such annotations are more distracting than helpful for users, and example code has different trade-offs than the implementation (first and foremost: demonstrate the library to the user).

**Comment by ChrisThrasher** [comment]
> Should we maybe exclude examples from this check?

That annotation is the only option we have for disabling this check in the examples short of entirely abandoning running clang-tidy on the examples. If we really care about setting a good example we could instead refactor those `main()`s to be less complex. That's a lot harder than slapping an annotation on top of `main()` but it means no annotation and an improved example program.

> Honestly, I don't really see how this reduces complexity. Sure, it removes an else and a level of indentation, but it also makes the two branches asymmetric and makes it less obvious that it's either-or.

The way I see it is that this check is promoting the use of [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html). In the case of only two options, a guard clause is semantically the same as a simple `if/else` but it's not uncommon to guard against more than one thing before you fall back to the default case. When there is more than one thing to guard against, it's a lot simpler to omit the `else` statements as much as possible than construct some big conditional logic tree. As a general rule, omitting the `else` is a good idea even if in select cases omitting that `else` isn't particularly helpful.

**Comment by Bromeon** [comment]
> If we really care about setting a good example we could instead refactor those `main()`s to be less complex.

Maybe there are some low-hanging gains to be had when extracting into a function. I would avoid taking this too far though, most important in the example is a somewhat linear code flow.

The problem with leaving the annotation is that it's not obvious to readers why this is needed (and in fact completely irrelevant since it's some SFML-internal CI config). It should at least be explained, but better left out.

---

> The way I see it is that this check is promoting the use of [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html). In the case of only two options, a guard clause is semantically the same as a simple `if`/`else` but it's not uncommon to guard against more than one thing before you fall back to the default case.



This starts to look like a good example where too much tooling gets in our way and takes away liberty in writing code. Guard clauses make sense in the example you linked, but also, it's a very academic example. Who writes code like this?
```js
function getPayAmount() {
  let result;
  if (isDead)
    result = deadAmount();
  else {
    if (isSeparated)
      result = separatedAmount();
    else {
      if (isRetired)
        result = retiredAmount();
      else
        result = normalPayAmount();
    }
  }
  return result;
}
```
Many people would do this instead:
```js
function getPayAmount() {
   if (isDead) 
      return deadAmount();
   else if (isSeparated)
      return separatedAmount();
   else if (isRetired)
      return retiredAmount();
   else
      return normalPayAmount();
}
```
And I don't see any problem with that: it's obvious that the 4 options are mutually exclusive. The "suggested" code is also cheating, because it has the expression on the same line; this would anyway not pass our formatter if it were C++.

Forbidding `if`/`else if`/`else` branches with `return` statements (one of the most basic language constructs) goes too far in my opinion.

We need to consider how many times contributors have to rewrite perfectly fine code just because a questionable rule says so. We are free to use guard clauses and recommend other code styles in review, but enforcing it is too much. We should focus on preventing bugs, not dictating code style to the letter; this can also turn into a frustrating experience for PR authors otherwise.

**Comment by Bromeon** [comment]
Maybe as another point that's not directly relevant to the argument in C++, but I thought might still be worth mentioning.

In many modern languages, there is a trend toward expression-based `if` statements, meaning you can do this:
```kotlin
// Kotlin
fun getPayAmount(): Amount {
   return if (isDead) 
      deadAmount()
   else if (isSeparated)
      separatedAmount()
   else if (isRetired)
      retiredAmount()
   else
      normalPayAmount()
}
```
```rust
// Rust (implicit return)
fn get_pay_amount() -> Amount {
    if is_dead() {
        dead_amount()
    } else if is_separated() {
        separated_amount()
    } else if is_retired() {
        retired_amount()
    } else {
        normal_pay_amount()
    }
}
```

**Comment by ChrisThrasher** [comment]
> The problem with leaving the annotation is that it's not obvious to readers why this is needed (and in fact completely irrelevant since it's some SFML-internal CI config). It should at least be explained, but better left out.

My impression is that users aren't looking at the examples all that much. I've never seen someone in the Discord allude to any particular example. That's why I don't think too deeply about a linter annotation in there. We already have some clang-format annotations in there anyways.

> This starts to look like a good example where too much tooling gets in our way and takes away liberty in writing code.

That's exactly my goal. I want to programmatically disallow writing code that I think is inferior to some given alternative. The liberty to write code with extra unnecessary control flow is not one I value.

> I need to read towards the end of each line to see the return, and with their "on-the-same-line" style all returns even have different indentation.

Not sure why you made a point about formatting when `return`s on the same line are not allowed by our clang-format config. Here is that same function with and without the unnecessary `else`s.

```js
function getPayAmount() {
   if (isDead) 
      return deadAmount();
   else if (isSeparated)
      return separatedAmount();
   else if (isRetired)
      return retiredAmount();
   else
      return normalPayAmount();
}
```

```js
function getPayAmount() {
   if (isDead) 
      return deadAmount();
   if (isSeparated)
      return separatedAmount();
   if (isRetired)
      return retiredAmount();
   return normalPayAmount();
}
```

> We need to consider how many times contributors have to rewrite perfectly fine code just because a questionable rule says so.

It's wonderful that our CI pipeline is getting stronger and strong so it does more of the work that previously had to happen in code review. I highly disagree with the "questionable" assessment.

**Comment by Bromeon** [comment]
> If we end up with a significant number of contributors annoyed that they're forced to simplify their code we can reevaluate, but I'm confident that won't happen.

You inverse the burden of proof here. Rather than having no one complain with the new setup, we should have some grounds to assume that the current setup is problematic.

---

> If you fundamentally question the value of cognitive complexity measurements then it's pointless to discuss ways to reduce it.

I don't, but we should focus on those that are less controversial and help prevent actual bugs. We should also not succumb to the fallacy "it's an automated tool, therefore it's right".

Forbidding `else` goes into "blindly following a tool for the sake of it" territory. There's no objective reason why that would improve anything; as I showed, there are downsides as well. In most of the codebases that I can remember, both styles coexisted, and I don't remember anyone ever pointing this out as a issue.

**TLDR: I think we're inventing problems here.**



---

> Anyways, let's not get too bogged down in `readability-else-after-return` since it's merely one technique for addressing cognitive complexity.

I agree, let's focus on the more productive metrics 🙂

**Comment by ChrisThrasher** [comment]
> You inverse the burden of proof here. Rather than having no one complain with the new setup, we should have some grounds to assume that the current setup is problematic.

https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

I already provided some positive evidence for this check.

> we should focus on those that are less controversial and help prevent actual bugs

Removing unnecessary code is a good step towards making bugs easier to see when reading code.

> We should also not succumb to the fallacy "it's an automated tool, therefore it's right".

I'm not. I'm aware that clang-tidy has some checks that are mostly neutral and only exist for those with particular preferences. [`readability-braces-around-statements`](https://clang.llvm.org/extra/clang-tidy/checks/readability/braces-around-statements.html) comes to mind. Note how that check does not rationalize itself. it merely states that it enforces a particular preference.

> Forbidding else goes into "blindly following a tool for the sake of it" territory.

Nothing blind about it. I have my rationale. You prefer extra `else`s. I think those extra `else`s obscure the code. This is our disagreement.

@eXpl0it3r Will have to do his BDFL duties on this one.

**Comment by Bromeon** [comment]
> Removing unnecessary code makes it easier to spot bugs. [...]
> I want to make the code more succinct. You prefer the verbosity of `else`s even when not necessary.

You word it as if `else` were "verbose" or "complex", whereas I see it as a fundamental language feature.

---

I'm not against using guard clauses where appropriate (e.g. early exit), but I'm against religiously following as if they _invariably_ led to better code. Which is absolutely not the case, so let's get to concrete examples instead of bikeshedding.

**Having `else` can prevent bugs.** To stay with your JS example, let's make one of the conditions have 2 subconditions:
```js
function getPayAmount() {
   if (isDead) {
      return deadAmount();
   } else if (isSeparated) {
      return separatedAmount();
   } else if (isRetired) {
      if (livesAtHome)
          return retiredAtHomeAmount();
      else
          return retiredAmount();
   } else {
      return normalPayAmount();
   }
}
```
Now I refactor it because we have a pandemic and everyone needs to stay at home:
```js
function getPayAmount()
{
   if (isDead) {
      return deadAmount();
   } else if (isSeparated) {
      return separatedAmount();
   } else if (isRetired) {
      if (livesAtHome)
          return retiredAtHomeAmount();
      // else - TODO re-enable later
      //    return retiredAmount();
   } else {
      return normalPayAmount();
   }
}
```
However, I forgot to handle one case. Now, the compiler can warn me that not all code paths have a return value.


Had I gone with your code instead:
```js
function getPayAmount() {
  if (isDead) return deadAmount();
  if (isSeparated) return separatedAmount();
  if (isRetired) {
     if (livesAtHome)
         return retiredAtHomeAmount();
     return retiredAmount();
  }
  return normalPayAmount();
}
```
After refactor:
```js
function getPayAmount() {
  if (isDead) return deadAmount();
  if (isSeparated) return separatedAmount();
  if (isRetired) {
     if (livesAtHome)
         return retiredAtHomeAmount();
     // return retiredAmount(); - TODO re-enable later
  }
  return normalPayAmount();
}
```
We now accidentally fall through to the `normalPayAmount()`, instead of getting a compiler warning.
In other words, a bug -- one that wouldn't have existed with `else`.

**Comment by ChrisThrasher** [comment]
I'll leave it here since we could talk in circles for hours more about this. Here are the two big values/philosophies motivating me. If you disagree with one or both then we will not agree on the value of outdenting these `else` blocks.

1. I value writing less code when possible. If I can remove code without affecting functionality, I'm going to do that 9 out of 10 times.

2. Static analyzers don't have to be perfect to be value. If a static analyzer pedantically insists on refactoring some code for no value but also finds the places where a refactor _would_ be valuable, that's a win. I'll put up with some annoyance if it finds the places where change is really valuable.

When I follow those two values, I come conclusions like "yes let's automatically outdent these unnecessary `else` blocks with this tool". I suspect you disagree somewhat with both values so we're not going to see eye to eye on this. Let's just wait for Exploiter to get around to reading this and he can make a decision based on whose values he aligns more with.

**Comment by ChrisThrasher** [comment]
I stand by this clang-tidy check as a good metric that SFML ought to add but also acknowledge that there isn't enough support to add this now so I'll close the issue. For now we can focus on more important API-affecting changes for SFML 3 like #2749 or #2766. In the future I hope we can reconsider this check since an automated tool for flagging overly complex functions is very useful to a large codebase like this.

**Comment by eXpl0it3r** [comment]
Chris and I had a brief chat on Discord, but I thought, I might put my response here as well.

I do think Cognitive Complexity can be a valuable metric, but it's in my opinion a bit of different category than other coding guidelines/standards, as it's something much more qualitative, where as other checks seem to also influence the actual safety of the code, making less indentations, is just a moving around of code to make things more readable.

As we do have some large portions of code that would require a lot of moving around of code, I don't really see the benefit of that investment right now. I wouldn't necessarily mind the *metric* itself, without enforcement of any kind, thus allowing us to consider improving things as we go, especially for say new code or areas we're touching.
